storage: improve batching algorithm when sending keys for GC

spencerkimball opened the pull request:
Previously the batching algorithm only considered the size of
top-level keys. This used to work well when we were doing
replica-evaluated Raft. With leader-evaluated, keys with many
versions can blow up into massive write batch values which can
end up clogging up Raft, causing performance degradation.

This PR augments the `GC.Filter` method to return the index of
the first version of a key to be GC'ed, allowing us to keep track
of the total number of keys to be GC'd. The GC queue is now able
to send a more appropriate number of keys per batch, which results
in significant improvements in 99th percentile latencies.

This was tested with a modified version of the kv load generator,
set up to write large keys (1024 bytes) with a short cycle length,
creating lots of key versions for GC. A three node local cluster
with the TTL set to 120 seconds could be made to experience up to
6 second 99th %ile latencies after about 10 minutes of running.
With this change, p99 latencies appear stable.

N.B. an alternative to this change was to let each replica handle
GC locally, and change the consistency checker to avoid considering
versions earlier than the GC threshold. However, this ran into
fairly tricky problems around keeping the MVCC stats values correct.

See #20052

Release note: improved p99 latencies for garbage collection of
previous versions of a key, when there are **many** versions.

spencerkimball committed:
storage: improve batching algorithm when sending keys for GC

Previously the batching algorithm only considered the size of
top-level keys. This used to work well when we were doing
replica-evaluated Raft. With leader-evaluated, keys with many
versions can blow up into massive write batch values which can
end up clogging up Raft, causing performance degradation.

This PR augments the `GC.Filter` method to return the index of
the first version of a key to be GC'ed, allowing us to keep track
of the total number of keys to be GC'd. The GC queue is now able
to send a more appropriate number of keys per batch, which results
in significant improvements in 99th percentile latencies.

This was tested with a modified version of the kv load generator,
set up to write large keys (1024 bytes) with a short cycle length,
creating lots of key versions for GC. A three node local cluster
with the TTL set to 120 seconds could be made to experience up to
6 second 99th %ile latencies after about 10 minutes of running.
With this change, p99 latencies appear stable.

N.B. an alternative to this change was to let each replica handle
GC locally, and change the consistency checker to avoid considering
versions earlier than the GC threshold. However, this ran into
fairly tricky problems around keeping the MVCC stats values correct.

See #20052

Release note (performance improvement): improved p99 latencies
for garbage collection of previous versions of a key, when there
are **many** versions.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20373)
<!-- Reviewable:end -->

tschottdorf commented:
Good point about the alternative suffering from stats complexities. Still possible, but certainly less appealing.
I think you could get the high latencies back if you used long keys, though, since you chunk on the count and not on the byte size. I think it's acceptable to hold off on that though.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 78 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-IzxWCJWObOXuKWQBM:-L-IzxWCJWObOXuKWQBN:b-l2nzg4) ([raw file](https://github.com/cockroachdb/cockroach/blob/022f85ae744a708ae4173855c96d8993073673a8/pkg/storage/gc_queue.go#L78)):*
> ```Go
> 
> 	// gcKeyVersionChunkSize is the default size for GCRequest's batch key size.
> 	gcKeyVersionChunkSize = 1000
> ```

Clarify that this counts versions (I'd even rename the const to `gcKeyVersionChunkCount` to make it obvious). I think the *right* thing to chunk on is key size though, as the old variable tried to do. For example, if you have 10mb keys (god forbid) you'd pile up 1000*10mb and would really want to send each key alone. OTOH, if you have 10k keys and they're all 1 byte each, no problem. (I think it's acceptable to hold off on that now, but it should be mentioned, say in a TODO).

---

*[pkg/storage/gc_queue.go, line 798 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-IwB1OJ5DGjDE5j7g1:-L-IwB1OJ5DGjDE5j7g2:bg3qg55) ([raw file](https://github.com/cockroachdb/cockroach/blob/022f85ae744a708ae4173855c96d8993073673a8/pkg/storage/gc_queue.go#L798)):*
> ```Go
> 						}
> 						txnMap[txnID] = txn
> 						infoMu.IntentTxns++
> ```

Only increment `IntentTxns` when you are first creating the entry in `txnMap`. The current code is overcounting.

---

*[pkg/storage/gc_queue.go, line 806 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-IzW-7k4bOroYZPkgn:-L-IzW-7k4bOroYZPkgo:bjdhxcl) ([raw file](https://github.com/cockroachdb/cockroach/blob/022f85ae744a708ae4173855c96d8993073673a8/pkg/storage/gc_queue.go#L806)):*
> ```Go
> 				}
> 				// See if any values may be GC'd.
> 				if idx, gcTS := gc.Filter(keys[startIdx:], vals[startIdx:]); gcTS != (hlc.Timestamp{}) {
> ```

Not really a request to change anything, but just wanted to point out that when there are many versions, we could've considered using an iterator here and jumping to the next key once we know the cutoff. This isn't possible any more after this change (since we need `len(keys[:startIdx])` to be known). I think that's ok.

---

*[pkg/storage/gc_queue.go, line 808 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-IwYp3DTXmyol24yLb:-L-IwYp3DTXmyol24yLc:b4d2q37) ([raw file](https://github.com/cockroachdb/cockroach/blob/022f85ae744a708ae4173855c96d8993073673a8/pkg/storage/gc_queue.go#L808)):*
> ```Go
> 				if idx, gcTS := gc.Filter(keys[startIdx:], vals[startIdx:]); gcTS != (hlc.Timestamp{}) {
> 					// Add key to slice of local keys to be deleted directly
> 					// from engine _after_ the GC threshold timestamp has been
> ```

Stale comment? We won't delete straight from the engine.

---

*[pkg/storage/gc_queue.go, line 812 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-Iww5rM9oTBVpzXBUG:-L-Iww5rM9oTBVpzXBUH:b-uzbi7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/022f85ae744a708ae4173855c96d8993073673a8/pkg/storage/gc_queue.go#L812)):*
> ```Go
> 					// GC requests through Raft.
> 					numVersions := int64(len(keys) - (startIdx + idx))
> 					if len(localGCKeys) > 0 && (localGCKeysVersionCount+numVersions) > gcKeyVersionChunkSize {
> ```

Why the `len(.)` check? 

---

*[pkg/storage/gc_queue.go, line 814 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-IxE58dZsTZ-0nRUK6:-L-IxE58dZsTZ-0nRUK7:be6tf2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/022f85ae744a708ae4173855c96d8993073673a8/pkg/storage/gc_queue.go#L814)):*
> ```Go
> 					if len(localGCKeys) > 0 && (localGCKeysVersionCount+numVersions) > gcKeyVersionChunkSize {
> 						gcKeys = append(gcKeys, localGCKeys)
> 						localGCKeys = []roachpb.GCRequest_GCKey{}
> ```

```
localGCKeys = nil
```

This code looks like it might benefit from use of a `bufalloc.ByteAllocator`, but not without measuring first (so not anytime soon).

---

*[pkg/storage/gc_queue_test.go, line 963 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-Iyop162ROMHPEpxQS:-L-Iyop162ROMHPEpxQT:bqctgpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/022f85ae744a708ae4173855c96d8993073673a8/pkg/storage/gc_queue_test.go#L963)):*
> // We wrote two batches worth of keys spread out, and two keys that each have enough old versions to fill a whole batch each, adding up to four batches in total.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-:-L-IvVLyKevar684_hBV:b-8v4omd)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reminder: add a [release note](https://github.com/cockroachdb/cockroach/blob/master/CONTRIBUTING.md)

spencerkimball commented:
I've switched to use byte count and added a release note.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 78 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-IzxWCJWObOXuKWQBM:-L-KU67xQPGF4VUkNNjs:b-s0o6rh) ([raw file](https://github.com/cockroachdb/cockroach/blob/022f85ae744a708ae4173855c96d8993073673a8/pkg/storage/gc_queue.go#L78)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Clarify that this counts versions (I'd even rename the const to `gcKeyVersionChunkCount` to make it obvious). I think the *right* thing to chunk on is key size though, as the old variable tried to do. For example, if you have 10mb keys (god forbid) you'd pile up 1000*10mb and would really want to send each key alone. OTOH, if you have 10k keys and they're all 1 byte each, no problem. (I think it's acceptable to hold off on that now, but it should be mentioned, say in a TODO).
</blockquote></details>

I've switched to using byte counts instead of number of keys.

---

*[pkg/storage/gc_queue.go, line 798 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-IwB1OJ5DGjDE5j7g1:-L-KUJKg8faL754Mr-Jp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/022f85ae744a708ae4173855c96d8993073673a8/pkg/storage/gc_queue.go#L798)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Only increment `IntentTxns` when you are first creating the entry in `txnMap`. The current code is overcounting.
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 806 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-IzW-7k4bOroYZPkgn:-L-KULnoE7N0OEXlmE7b:bqr852z) ([raw file](https://github.com/cockroachdb/cockroach/blob/022f85ae744a708ae4173855c96d8993073673a8/pkg/storage/gc_queue.go#L806)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Not really a request to change anything, but just wanted to point out that when there are many versions, we could've considered using an iterator here and jumping to the next key once we know the cutoff. This isn't possible any more after this change (since we need `len(keys[:startIdx])` to be known). I think that's ok.
</blockquote></details>

I thought about that a lot actually. The idea originally was to allow an open ended policy for the `GC.Filter` function which could consider all values together. So far we haven't used it for anything, so might be better to punt. Worst case we could pass the iterator to the filter function instead so it could do what it needed to do. Things seem reasonably performant in any case. I'm going to leave as is.

---

*[pkg/storage/gc_queue.go, line 808 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-IwYp3DTXmyol24yLb:-L-KUmjFa8GFCdwY7L32:b-7tw5gx) ([raw file](https://github.com/cockroachdb/cockroach/blob/022f85ae744a708ae4173855c96d8993073673a8/pkg/storage/gc_queue.go#L808)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Stale comment? We won't delete straight from the engine.
</blockquote></details>

Yeah, that was left over from original direction. Removed.

---

*[pkg/storage/gc_queue.go, line 812 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-Iww5rM9oTBVpzXBUG:-L-KUxok7GnUsC2wgfYc:b-ocvuzx) ([raw file](https://github.com/cockroachdb/cockroach/blob/022f85ae744a708ae4173855c96d8993073673a8/pkg/storage/gc_queue.go#L812)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Why the `len(.)` check? 
</blockquote></details>

In case the total count (or now size) of all of the versions on the very first key exceeds the threshold. In that case, we'd append an empty gcKeys slice.

---

*[pkg/storage/gc_queue.go, line 814 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-IxE58dZsTZ-0nRUK6:-L-KVDpngOczbsoGb1Uz:b-edy903) ([raw file](https://github.com/cockroachdb/cockroach/blob/022f85ae744a708ae4173855c96d8993073673a8/pkg/storage/gc_queue.go#L814)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

```
localGCKeys = nil
```

This code looks like it might benefit from use of a `bufalloc.ByteAllocator`, but not without measuring first (so not anytime soon).
</blockquote></details>

Not sure what that's for...you mean just to avoid reallocating large slices?

---

*[pkg/storage/gc_queue_test.go, line 963 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-Iyop162ROMHPEpxQS:-L-KVUe-NRK7idfDAKO6:b-uldnpx) ([raw file](https://github.com/cockroachdb/cockroach/blob/022f85ae744a708ae4173855c96d8993073673a8/pkg/storage/gc_queue_test.go#L963)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

> // We wrote two batches worth of keys spread out, and two keys that each have enough old versions to fill a whole batch each, adding up to four batches in total.
</blockquote></details>

Good add.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-:-L-KSiEWf5RVkdsMPrAl:b-gsdhjj)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Update the commit message (byte count). Also, the release note should [have a category](https://github.com/cockroachdb/cockroach/blob/master/CONTRIBUTING.md#commit-messages). I think in this case

> Release note (performance improvement): ...

is appropriate.

This change is starting to look really good! I had one comment which I think substantially improves on this, namely being able to chunk single keys with lots of versions by creating a sequence of GCRequests for the key with ascending target timestamp. Looks like a big win (for example, would occur as folks tighten the GC TTL on overwrite-heavy workloads) and the code is basically there already in your current work.

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 78 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-IzxWCJWObOXuKWQBM:-L-SDad6uQY1jIon_Cls:bna6pzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/022f85ae744a708ae4173855c96d8993073673a8/pkg/storage/gc_queue.go#L78)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

I've switched to using byte counts instead of number of keys.
</blockquote></details>

Rename to `gcKeyVersionChunkBytes`

---

*[pkg/storage/gc_queue.go, line 814 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-IxE58dZsTZ-0nRUK6:-L-SFguR_-Qtf0o_h-A6:bb6r27h) ([raw file](https://github.com/cockroachdb/cockroach/blob/022f85ae744a708ae4173855c96d8993073673a8/pkg/storage/gc_queue.go#L814)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Not sure what that's for...you mean just to avoid reallocating large slices?
</blockquote></details>

Yeah, it could reduce the allocations for getting all these key values. But not for now.

---

*[pkg/storage/gc_queue.go, line 578 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SE5-YJAT3PRwz6cae:-L-SE5-ZQvZ2qBJrYunN:b-7c9qho) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue.go#L578)):*
We don't actually want to put any data into the first request. The first one is special in that it sets the GC thresholds, and as such, it can't intermingle with any on-going reads (that is, it's pretty heavy on the command queue). Consequently, we want it as small as possible so that it slips through Raft quickly. This may be less of an issue now that we're properly chunking, but at the very least a table with ~1mb keys would have that problem, so it's worth the one-off.

While you're here, please add that as a comment. For example,

> The first request is intentionally kept very small since it updates the Range-wide GCThresholds, and thus must block all reads and writes while it is being applied.

---

*[pkg/storage/gc_queue.go, line 800 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SFH2MxcU42WOZR1m9:-L-SFH2MxcU42WOZR1mA:b-vh0utn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue.go#L800)):*
> ```Go
> 							infoMu.IntentTxns++
> 						}
> 						txnMap[txnID] = txn
> ```

Move this into the `!ok` along with `txn := ...` above.

---

*[pkg/storage/gc_queue.go, line 816 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SGNsUvTflPhrNVSAX:-L-SGNsUvTflPhrNVSAY:b-lwn50u) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue.go#L816)):*
> // If the current chunk is non-empty and the current key would bring it over the target size, flush and start a new chunk.

---

*[pkg/storage/gc_queue.go, line 821 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SGt-uSc-ELs9pgNQN:-L-SGt-uSc-ELs9pgNQO:b-gpe484) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue.go#L821)):*
> ```Go
> 						batchGCKeysByteCount = 0
> 					}
> 					batchGCKeys = append(batchGCKeys, roachpb.GCRequest_GCKey{Key: expBaseKey, Timestamp: gcTS})
> ```

So if you had alternatingly 1kb and 1mb keys to GC (not unlikely with column families), it would be pretty inefficient because you'd be sending `1kb` and `1mb` chunks. The straightforward alternative would be adding to a chunk and resetting after it goes over target size, but then we might go over by quite a bit. I think what you're doing here is probably better. There's of course also the third option, which is keeping the chunks in a min-heap (by size) and filling them up optimally, but pretty sure that's a tangent we don't have to go on right now.

But I'm still worried about single keys with a huge history, which you still won't chunk down here. But, what you could do, is artificially lower the GC timestamp so that it touches less version, and chunk on that. That is, you have a key with versions at 1000, 999, ..., 1 and each version is 8kb (and say your target size is 80kb) you'd emit GC requests for `ts <= 10, ts <= 20, ... ts <= 1000` because evaluated in that order, each of them will generate ~80kb of deletions.

Your current code has all the ingredients. You just need to change the loop over `keys` to go in reverse direction (old versions to new) and move the chunk-emitting logic into it. I think that's worth doing and also means that chunks will be more balanced in general, except when folks throw very large keys into the mix (at which point, not much we can do except damage control).

---

*[pkg/storage/gc_queue.go, line 822 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SGD_2SIVc86eqSddS:-L-SGD_2SIVc86eqSddT:b6kjbty) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue.go#L822)):*
> ```Go
> 					}
> 					batchGCKeys = append(batchGCKeys, roachpb.GCRequest_GCKey{Key: expBaseKey, Timestamp: gcTS})
> 					batchGCKeysByteCount += byteCount
> ```

You can move this line up above the `if len(.)` and then just use `batchGCKeysByteCount` in the `if`.

---

*[pkg/storage/gc_queue_test.go, line 917 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SJzynefZBNoVlXyWB:-L-SJzynefZBNoVlXyWC:bjh8w1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue_test.go#L917)):*
> ```Go
> 	tc.StartWithStoreConfig(t, stopper, tsc)
> 
> 	keyCount := 1000
> ```

nit: `const keyCount = 1000`

---

*[pkg/storage/gc_queue_test.go, line 919 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SK0A_aj5mwhlTJ0dm:-L-SK0A_aj5mwhlTJ0dn:b-pi5chs) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue_test.go#L919)):*
> ```Go
> 	keyCount := 1000
> 	if gcKeyVersionChunkSize%keyCount != 0 {
> 		panic("expected gcKeyVersionChunkSize to be a multiple of 1000")
> ```

`t.Fatalf("expected GC chunk size to be a multiple of %s", keyCount)

---

*[pkg/storage/gc_queue_test.go, line 921 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SKHQhSSDd3OjVvVjC:-L-SKHQhSSDd3OjVvVjD:bjav31i) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue_test.go#L921)):*
> ```Go
> 		panic("expected gcKeyVersionChunkSize to be a multiple of 1000")
> 	}
> 	keySize := gcKeyVersionChunkSize / keyCount
> ```

nit: can be a `const` too, I think! Always nice to see at first glance the things that can't change.

---

*[pkg/storage/gc_queue_test.go, line 923 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SJsN36ld7fTMQQ0wh:-L-SJsN36ld7fTMQQ0wi:b-eanox) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue_test.go#L923)):*
> // length `keySize`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-:-L-SDOgngET8_cnwRci7:by4fhcj)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
OK, I've incorporated that change (and further expanded the unittest).

---

Review status: 2 of 4 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/storage/gc_queue.go, line 78 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-IzxWCJWObOXuKWQBM:-L-TUdYOTsbZboToTPTg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/022f85ae744a708ae4173855c96d8993073673a8/pkg/storage/gc_queue.go#L78)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Rename to `gcKeyVersionChunkBytes`
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 814 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-IxE58dZsTZ-0nRUK6:-L-TUvEnEiwlIdBXJ___:b2nhb2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/022f85ae744a708ae4173855c96d8993073673a8/pkg/storage/gc_queue.go#L814)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Yeah, it could reduce the allocations for getting all these key values. But not for now.
</blockquote></details>

Ack

---

*[pkg/storage/gc_queue.go, line 578 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SE5-YJAT3PRwz6cae:-L-TUmXv8W1YuWJHJGoX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue.go#L578)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

We don't actually want to put any data into the first request. The first one is special in that it sets the GC thresholds, and as such, it can't intermingle with any on-going reads (that is, it's pretty heavy on the command queue). Consequently, we want it as small as possible so that it slips through Raft quickly. This may be less of an issue now that we're properly chunking, but at the very least a table with ~1mb keys would have that problem, so it's worth the one-off.

While you're here, please add that as a comment. For example,

> The first request is intentionally kept very small since it updates the Range-wide GCThresholds, and thus must block all reads and writes while it is being applied.
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 800 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SFH2MxcU42WOZR1m9:-L-TUtepdM6nl1ehb7YJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue.go#L800)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Move this into the `!ok` along with `txn := ...` above.
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 816 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SGNsUvTflPhrNVSAX:-L-TV5n9y4KaXCqt4UGn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue.go#L816)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

> // If the current chunk is non-empty and the current key would bring it over the target size, flush and start a new chunk.
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 821 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SGt-uSc-ELs9pgNQN:-L-TgXlcno1Sr3tTM_C4:b-8l7dx6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue.go#L821)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

So if you had alternatingly 1kb and 1mb keys to GC (not unlikely with column families), it would be pretty inefficient because you'd be sending `1kb` and `1mb` chunks. The straightforward alternative would be adding to a chunk and resetting after it goes over target size, but then we might go over by quite a bit. I think what you're doing here is probably better. There's of course also the third option, which is keeping the chunks in a min-heap (by size) and filling them up optimally, but pretty sure that's a tangent we don't have to go on right now.

But I'm still worried about single keys with a huge history, which you still won't chunk down here. But, what you could do, is artificially lower the GC timestamp so that it touches less version, and chunk on that. That is, you have a key with versions at 1000, 999, ..., 1 and each version is 8kb (and say your target size is 80kb) you'd emit GC requests for `ts <= 10, ts <= 20, ... ts <= 1000` because evaluated in that order, each of them will generate ~80kb of deletions.

Your current code has all the ingredients. You just need to change the loop over `keys` to go in reverse direction (old versions to new) and move the chunk-emitting logic into it. I think that's worth doing and also means that chunks will be more balanced in general, except when folks throw very large keys into the mix (at which point, not much we can do except damage control).
</blockquote></details>

Swell idea! Implemented.

---

*[pkg/storage/gc_queue.go, line 822 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SGD_2SIVc86eqSddS:-L-TVP9EUWp3YfdzTp62:b-o24k2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue.go#L822)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

You can move this line up above the `if len(.)` and then just use `batchGCKeysByteCount` in the `if`.
</blockquote></details>

Then I'd have to reset `batchGCKeysByteCount` to `byteCount` if the `if` clause is true, which I find less readable.

---

*[pkg/storage/gc_queue_test.go, line 917 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SJzynefZBNoVlXyWB:-L-TVpqKQVIIrHlDG1gh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue_test.go#L917)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

nit: `const keyCount = 1000`
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue_test.go, line 919 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SK0A_aj5mwhlTJ0dm:-L-TVv0w_8YmWMZxKMeF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue_test.go#L919)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

`t.Fatalf("expected GC chunk size to be a multiple of %s", keyCount)
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue_test.go, line 921 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SKHQhSSDd3OjVvVjC:-L-TW2YyVsukzbTZKr9y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue_test.go#L921)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

nit: can be a `const` too, I think! Always nice to see at first glance the things that can't change.
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue_test.go, line 923 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SJsN36ld7fTMQQ0wh:-L-TW63BQsSbboRjbHcQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue_test.go#L923)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

> // length `keySize`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-:-L-TTsP2QRAR7zQQP5nz:b-j40q93)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> :tada: 

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/gc_queue.go, line 800 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SFH2MxcU42WOZR1m9:-L-UB0bc5G0FBgbTS0ma:b3dwcdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue.go#L800)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

and the `txn :=` is better in the `if _, ok` branch too. Actually you can just inline the def: `txnMap[txnID] = &roachpb.Transaction { ... }`

---

*[pkg/storage/gc_queue.go, line 79 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-UAGTqwk1NOVFILmec:-L-UAGTqwk1NOVFILmed:bc3u2y9) ([raw file](https://github.com/cockroachdb/cockroach/blob/af0ca7d42db85a78ea50cc352b0c53e4544aaa5d/pkg/storage/gc_queue.go#L79)):*
> ```Go
> 	// gcKeyVersionChunkBytes is the threshold size for splitting
> 	// GCRequests into multiple batches.
> 	gcKeyVersionChunkBytes = 1024 * 1000
> ```

1mb could still show up significantly in latencies. Might be worth starting smaller here.  How about 256 KiB (i.e. `1000*256`)?

---

*[pkg/storage/gc_queue.go, line 570 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-UAnCqYL5qV5TXnCz7:-L-UAnCqYL5qV5TXnCz8:b-tpc4xd) ([raw file](https://github.com/cockroachdb/cockroach/blob/af0ca7d42db85a78ea50cc352b0c53e4544aaa5d/pkg/storage/gc_queue.go#L570)):*
> ```Go
> ) []roachpb.GCRequest {
> 	var template roachpb.GCRequest
> 	var ret []roachpb.GCRequest
> ```

nit: `ret := make([]roachpb.GCRequest, 0, len(gcKeys)+1)`

---

*[pkg/storage/gc_queue.go, line 818 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-UBWFFM9cUMOED6QkP:-L-UBWFFM9cUMOED6QkQ:bsf6asq) ([raw file](https://github.com/cockroachdb/cockroach/blob/af0ca7d42db85a78ea50cc352b0c53e4544aaa5d/pkg/storage/gc_queue.go#L818)):*
> // size, add the current timestamp and start a new chunk.

(if my reading of this code is correct, this is how it works now, where before you'd save the current key for the next batch).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-:-L-UGb-ZPTH3tZwOmAI9:b-5esoy1)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 3 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/storage/gc_queue.go, line 800 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-SFH2MxcU42WOZR1m9:-L-X2dpvepjXZgfuJFNr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e89d6b88bdd750180696747fbb549a3088d80f7/pkg/storage/gc_queue.go#L800)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

and the `txn :=` is better in the `if _, ok` branch too. Actually you can just inline the def: `txnMap[txnID] = &roachpb.Transaction { ... }`
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 79 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-UAGTqwk1NOVFILmec:-L-X2BAIoXRzmyYBajn3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af0ca7d42db85a78ea50cc352b0c53e4544aaa5d/pkg/storage/gc_queue.go#L79)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

1mb could still show up significantly in latencies. Might be worth starting smaller here.  How about 256 KiB (i.e. `1000*256`)?
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 570 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-UAnCqYL5qV5TXnCz7:-L-X2RFgsKsO5WeqEaES:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af0ca7d42db85a78ea50cc352b0c53e4544aaa5d/pkg/storage/gc_queue.go#L570)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

nit: `ret := make([]roachpb.GCRequest, 0, len(gcKeys)+1)`
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 818 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373#-L-UBWFFM9cUMOED6QkP:-L-X43Q7AeOcGov1oqPe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af0ca7d42db85a78ea50cc352b0c53e4544aaa5d/pkg/storage/gc_queue.go#L818)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

> // size, add the current timestamp and start a new chunk.

(if my reading of this code is correct, this is how it works now, where before you'd save the current key for the next batch).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20373)*
<!-- Sent from Reviewable.io -->
